Provide custom routes to avoid requiring admin privileges.

whikloj opened the pull request:
Currently this module has a requirement of "administer site configuration" to access the 3 routes provided by it. It also sets up some other routes that require "administer field display" and (as far as I can tell) "Content: administer fields" permissions to get all the various field information via the REST module.

This PR will:
* add a new `use islandora workbench` permission
* add 3 new routes to provide the same information as 3 of the routes provided by the REST module
* allow people with the new permission to access all 6 routes provided by this module.

These routes will require a change in islandora_workbench to know to call them, PR coming. Also I have not touched all the various endpoints used (i.e. creating taxonomy terms) so this might be the first of multiple steps to not needing admin privileges.

Lastly, I did update the permissions on the views this module creates...maybe that is not desirable? I can revert that change and users will just have to go into their views and update them.

whikloj commented:
Oops this is related to #24

mjordan commented:
WGT the role config above, we can add it to the Integration module so it creates it automatically, or share the config and document how to create it.

whikloj commented:
Requiring all the "administer" permissions makes this PR redundant. I'll need have to run the integration tests to see what I'm missing. Is there a docker-compose.yml you use to set up the test infrastructure?

mjordan commented:
I test against a standard ISLE starter site, which is also used in the Github CI workflow.

Let me do a bit more testing to confirm that the "administer" items are necessary. I'll try to do this ASAP so we can keep the momentum going.

whikloj commented:
I'm out this week, but I'm guessing part of the issue is instead of giving these administer permissions you have to give individual permissions to each content, media type and taxonomy you want to use. I'll have to pull my test user to see the full breath of permissions

mjordan commented:
Testing now. I don't know why (most likely user error of some sort) but all of the node administer permissions are unnecessary.  I'm moving on to media next. Will let you know how it goes.

mjordan commented:
Some progress:

I have been able to remove all of the "administer" permissions other than `administer taxonomy`. Without it, tests reveal the following:

`WARNING - Term 'Kirk, James T.' not created, HTTP response code was 403, response body was {"message":"The following permissions are required: \u0027create terms in person\u0027 OR \u0027administer taxonomy\u0027."}.`

One caveat here is that integration test coverage for creating media or terms that have their own fields other than `title` is incomplete. I think some proxy routes for those entity types similar to the ones you've created for nodes might let us use a profile config like:

```
uuid: e8369567-bd4c-425c-b6c0-0256f393cd8d
langcode: en
status: true
dependencies:
  config:
    - node.type.islandora_object
  module:
    - islandora
    - islandora_workbench_integration
    - media
    - node
_core:
  default_config_hash: e6W733rBSNmewsIQsvWQ-8mIE5H-CIeArLBwIbN5vVQ
id: workbenchusers
label: 'Workbench user'
weight: 3
is_admin: null
permissions:
  - 'create islandora_object content'
  - 'create media'
  - 'delete any islandora_object content'
  - 'delete any media'
  - 'delete islandora_object revisions'
  - 'delete own islandora_object content'
  - 'edit own islandora_object content'
  - 'manage media'
  - 'manage members'
  - 'revert islandora_object revisions'
  - 'update any media'
  - 'use islandora workbench'
  - 'view checksums'
  - 'view islandora_object revisions'
  - 'view media'
```

mjordan commented:
@whikloj the issue I have with roles that contain so many specific permissions is that they will not be portable. But, I realize that there is an inverse relationship between portability and general privilege levels. I guess the best we can do is offer a default role that covers the core Islandora vocabularies accompanied by guidance on how to modify the role locally to account for custom vocabularies that aren't present in the default role. I wonder if we can also provide a script or a new workbench task that can test what a local admin will need to do to the default role, should they want to use that role or a copy of it locally.

mjordan commented:
> I wonder if we could build it into the integration module to check the permissions of a role against the known permissions needed and list those missing.

Yes, that would be the best place for it.

whikloj commented:
Ok. I was able to run all the tests and have them pass. I had to give my test user the Manage members permission (not sure why I hadn't done that before)

I also had to add a couple route checks and expose a file upload permission from the rest module so we could give it to people. 

I also added a multiple permissions plugin for Views to allow people to have either the old permissions or the new one and updated the view config.

There is a third view that users need access to, that is the "terms_in_vocabulary" one. I'm not sure where is comes from.

I also have not worked on the permissions check yet.

whikloj commented:
Ok, I think I am done here. 

I may play around with adding a test role with the `'use islandora workbench'` permission to the Islandora instance stood up in the `islandora_workbench` test infrastructure. But that won't be helpful right now.

whikloj commented:
Okay, sorry for the delay. I noticed the actions are not running here, but my fork was running them and complaining about linting and test failures. I have fixed those.

whikloj commented:
Just noting that tests for this pass. https://github.com/whikloj/islandora_workbench_integration/actions/runs/17311445980

bondjimbond commented:
@mjordan Is this ready to merge?

mjordan commented:
Nope, I'm just back from vacation and will need to catch up on all the work @whikloj did in the last little while. I'll try to find some time to get set up for testing this in the next day or so.

whikloj commented:
Actually, hold on this. I hit a bug where my MediaViewAccess is causing a WSOD.

whikloj commented:
Hi @bondjimbond, my branch is not published to packagist so you would have to pull the git repository. Essentially, removing the old `islandora_workbench_integration` directory from your Drupal's `web/modules/contrib` directory and then pull in either the PR branch or my permissions branch. 

If this is a development instance then ok, but otherwise you might want to wait rather than mess with your setup.

bondjimbond commented:
Yeah, I guess it's best to wait. But you feel this is ready to go?

whikloj commented:
That is a hard question to answer definitively. Workbench does so many things and we are just starting our use of it. So I can say I have used it for the "create" task and it passes the tests in `islandora_workbench` and allows users to perform the same test actions as an admin level user.
